src/cargo/core/global_cache_tracker.rs

```diff
@@ -0,0 +1,1754 @@
+                let encoded_git_name = row.get_unwrap(0);
+                let short_name = row.get_unwrap(1);
+                let size = row.get_unwrap(2);
+                let timestamp = row.get_unwrap(3);
+                let kind = GitCheckout {
+                    encoded_git_name,
+                    short_name,
+                    size,
+                };
+                Ok((kind, timestamp))
+            })?
+            .collect::<Result<Vec<_>, _>>()?;
+        Ok(rows)
+    }
+
+    /// Returns whether or not an auto GC should be performed, compared to the
+    /// last time it was recorded in the database.
+    pub fn should_run_auto_gc(&mut self, frequency: Duration) -> CargoResult<bool> {
+        trace!("should_run_auto_gc");
+        if self.auto_gc_checked_this_session {
+            return Ok(false);
+        }
+        let last_auto_gc: Timestamp =
+            self.conn
+                .query_row("SELECT last_auto_gc FROM global_data", [], |row| row.get(0))?;
+        let should_run = last_auto_gc + frequency.as_secs() < now();
+        trace!(
+            "last auto gc was {}, {}",
+            last_auto_gc,
+            if should_run { "running" } else { "skipping" }
+        );
+        self.auto_gc_checked_this_session = true;
+        Ok(should_run)
+    }
+
+    /// Writes to the database to indicate that an automatic GC has just been
+    /// completed.
+    pub fn set_last_auto_gc(&self) -> CargoResult<()> {
+        self.conn
+            .execute("UPDATE global_data SET last_auto_gc = ?1", [now()])?;
+        Ok(())
+    }
+
+    /// Deletes files from the global cache based on the given options.
+    pub fn clean(&mut self, clean_ctx: &mut CleanContext<'_>, gc_opts: &GcOpts) -> CargoResult<()> {
+        self.clean_inner(clean_ctx, gc_opts)
+            .with_context(|| "failed to clean entries from the global cache")
+    }
+
+    fn clean_inner(
+        &mut self,
+        clean_ctx: &mut CleanContext<'_>,
+        gc_opts: &GcOpts,
+    ) -> CargoResult<()> {
+        let _p = crate::util::profile::start("cleaning global cache files");
+        let config = clean_ctx.config;
+        let base_git_path = config.git_path().into_path_unlocked();
+        let base = BasePaths {
+            index: config.registry_index_path().into_path_unlocked(),
+            git_db: base_git_path.join("db"),
+            git_co: base_git_path.join("checkouts"),
+            crate_dir: config.registry_cache_path().into_path_unlocked(),
+            src: config.registry_source_path().into_path_unlocked(),
+        };
+        let now = now();
+        trace!("cleaning {gc_opts:?}");
+        let tx = self.conn.transaction()?;
+        let mut delete_paths = Vec::new();
+        // This can be an expensive operation, so only perform it if necessary.
+        if gc_opts.is_download_cache_opt_set() {
+            // TODO: Investigate how slow this might be.
+            Self::sync_db_with_files(
```

Comment:
`is_download_cache_opt_set` takes into account settings from auto-gc but I thought the sync isn't supposed to happen on auto-gc due to how slow it is?

See also https://github.com/rust-lang/cargo/pull/12634#discussion_r1320310008

Reply:
Yea, the comment on `sync_db_with_files` was out-of-date. It always runs the sync, which is not too terribly slow. The slow part is the `size` scanning, which should only be done if requested in a manual gc.